### PR TITLE
BioBlend development tutorial

### PR DESCRIPTION
**Pull request by rikeshi**
New tutorial: Contributing to BioBlend as a developer

### CODE REVIEW

**Comment by hexylena** [comment]
Hi @rikeshi, wow, looks like a fantastic start! So nice to have this content in tutorial format, I would've liked this when I was starting out with bioblend.

We'll need to make some formatting changes to bring this in line with GTN standards. Is that ok for you?

**Comment by rikeshi** [comment]
Yep, that's okay with me. I'm open to suggestions. Or if you want to make changes directly, go ahead.

**Comment by hexylena** [comment]
Oh that'd be great. The main thing is we should use more hands-on boxes. We use those for any time the user needs to do something, run some command, add code to a file, etc. There are lots of examples in the admin tutorials, e.g. https://raw.githubusercontent.com/galaxyproject/training-material/main/topics/admin/tutorials/ansible-galaxy/tutorial.md 

```
> ### {% icon hands_on %} Hands-on: Configuration files
>
> 1. Create a `ansible.cfg` file (next to your playbook) to [configure settings](https://docs.ansible.com/ansible/2.9/reference_appendices/config.html) like the inventory file (and save ourselves some typing!), or the Python interpreter to use:
>
>    ```ini
>    [defaults]
>    interpreter_python = /usr/bin/python3
>    inventory = hosts
>    retry_files_enabled = false
>    ```
>
>    > ### {% icon tip %} CentOS7
>    > As mentioned in the "Ubuntu or Debian, CentOS or RHEL?" comment above, if you are using CentOS7 do not set `interpreter_python` in `ansible.cfg` .
>    {: .tip}
>
>    There is an additional useful option that you might want to add to your `ansible.cfg` file if you are connecting over SSH:
>
>    ```ini
>    [ssh_connection]
>    pipelining = true
>    ```
>
>    Pipelining will make [Ansible run faster](https://docs.ansible.com/ansible/2.9/reference_appendices/config.html#ansible-pipelining) by significantly reducing the number of new SSH connections that must be opened.
>
> 2. Create the `hosts` inventory file if you have not done so yet, defining a `[galaxyservers]` group with the address of the host where you want to install Galaxy. If you are running Ansible on the same machine where Galaxy will be installed to, you should set the `ansible_connection=local` variable. Lastly, you should explicitly set the `ansible_user` variable to the username to use when connecting to the server. Ansible has changed its behaviour over time regarding whether or not `ansible_user` is defined, and it is most effective to define it explicitly even when it can sometimes be inferred.
>
>    > > ### {% icon code-in %} Input: Bash
>    > > ```bash
>    > > cat hosts
>    > > ```
>    > {: .code-in}
>    >
>    > > ### {% icon code-out %} Output: Bash
>    > >
>    > > Your hostname is probably different:
>    > >
>    > > ```ini
>    > > [galaxyservers]
>    > > gat-88.training.galaxyproject.eu ansible_connection=local ansible_user=ubuntu
>    > > ```
>    > {: .code-out}
>    {: .code-2col}
>
{: .hands_on}
```

I think some of your sections could be collapsed there (e.g. the writing the python function with documentation/etc.) could be one hands on box.

**Comment by rikeshi** [comment]
Thanks @simonbray !

I made a start with the hands-on boxes. Not sure if you have write access. I added you as a collaborator to my fork, so now you should be able to edit either way.

Inside the boxes, paragraph breaks seem to not be displayed with additional white space. This makes the text look really dense, which I don't think is very nice. But I do not know how to fix this.

**Comment by hexylena** [comment]
> Inside the boxes, paragraph breaks seem to not be displayed with additional white space. This makes the text look really dense, which I don't think is very nice. But I do not know how to fix this.

No, that's been a longstanding GTN issue. I'll look into fixing it!

**Comment by hexylena** [comment]
Fixed the whitespace, it looks a lot nicer now.

**Review by hexylena — DISMISSED** [review]

**Comment by hexylena** [comment]
This looks great, I'm happy to merge this as is! Fantastic work y'all

**Comment by hexylena** [comment]
We could also move the tips into proper 'tip' boxes? It looks like the 'tips' section at the end is only referenced once, and having them as (collapsed by default) tips next to the place where it's discussed could be a lot easier for students to read the tip, and remain where they are in the lesson, rather than having to navigate back to where they should be.

**Comment by simonbray** [comment]
Let me have a read through first - I will have a look at the tip boxes as well

**Comment by hexylena** [comment]
Sounds good @simonbray, I've assigned it to you. Ping me when you're ready to merge!

File: topics/dev/tutorials/bioblend-dev/tutorial.md
Comment (rikeshi) [inline topics/dev/tutorials/bioblend-dev/tutorial.md:203]:
```suggestion
>    > ### {% icon code-out %} Output: Example JSON response
>    > ```javascript
>    > [{
>    >     'history_id': '2f94e8ae9edff68a',
>    >     'id': 'df7a1f0c02a5b08e',
>    >     'model_class': 'WorkflowInvocation',
>    >     'state': 'new',
>    >     'update_time': '2015-10-31T22:00:22',
>    >     'uuid': 'c8aa2b1c-801a-11e5-a9e5-8ca98228593c',
>    >     'workflow_id': '03501d7626bd192f',
>    > }]
>    > ```
>    >
>    > The above response would mean that our user only once invoked a workflow. The workflow's encoded ID is listed, as well as the encoded ID of the history containing this workflow. The state `new` indicates that the request to invoke this workflow is still pending and that the invocation has not been scheduled yet.
>    {:.code-out}
```

File: topics/dev/tutorials/bioblend-dev/tutorial.md
Comment (rikeshi) [inline topics/dev/tutorials/bioblend-dev/tutorial.md:235]:
```suggestion
>    > ### {% icon code-in %} Input: curl
>    > In this example we include two query parameters in our request, `limit` and `include_terminal`.
>    >
>    > ```shell
>    > curl -X GET -H 'x-api-key: <API_KEY>' \
>    >      'localhost:8080/api/invocations?limit=5&include_terminal=False'
>    > ```
>    >
>    > With `limit=5` we limit the result to at most 5 invocations. With `include_terminal=False` we indicate that invocations in a terminal state should be excluded from the results.
>    {:.code-in}
```

File: topics/dev/tutorials/bioblend-dev/tutorial.md
Comment (rikeshi) [inline topics/dev/tutorials/bioblend-dev/tutorial.md:262]:
```suggestion
>    > ### {% icon code-in %} Input: curl
>    > ```shell
>    > curl -X POST \
>    >     -H 'x-api-key: <API_KEY>' \
>    >     -H 'Content-Type: application/json' \
>    >     -d '{"name":"MyNewHistory"}' \
>    >     'localhost:8080/api/histories'
>    > ```
>    >
>    > This creates a new history with the name ‘MyNewHistory’.
>    {:.code-in}
```

**Comment by rikeshi** [comment]
Thanks for improving the formatting @hexylena !

**Review by rikeshi — COMMENTED** [review]
Just noticed the `> >`, which are unnecessary here I think. I'll stick with suggestions since @simonbray is assigned.

**Review by simonbray — APPROVED** [review]

**Comment by simonbray** [comment]
@hexylena ready to merge if you are

**Review by hexylena — APPROVED** [review]
This looks fantastic @rikeshi! Thanks for the fixes @simonbray they look good as well. The only change I've made is to revert it to a definition list since that's the most semantically correct structure for defining/explaining that long list of terms.

I'm going to approve and merge this!

**Comment by simonbray** [comment]
Thank you @rikeshi! Really nice work. :)

**Pull request merged**
